pkg/event/tigris.go

```diff
@@ -0,0 +1,218 @@
+package event
+
+import (
+	"context"
+
+	"github.com/tigrisdata/tigris-client-go/fields"
+	"github.com/tigrisdata/tigris-client-go/filter"
+	"github.com/tigrisdata/tigris-client-go/sort"
+	"github.com/tigrisdata/tigris-client-go/tigris"
+	"github.com/warrant-dev/warrant/pkg/database"
+	"github.com/warrant-dev/warrant/pkg/service"
+)
+
+type TigrisRepository struct {
+	db *tigris.Database
+}
+
+func NewTigrisRepository(db *database.Tigris) (TigrisRepository, error) {
+	tdb, err := db.T.OpenDatabase(context.TODO(), &ResourceEvent{}, &AccessEvent{})
+	if err != nil {
+		return TigrisRepository{}, err
+	}
+
+	return TigrisRepository{
+		db: tdb,
+	}, nil
+}
+
+func (repo TigrisRepository) TrackResourceEvent(ctx context.Context, resourceEvent ResourceEventModel) error {
+	return repo.TrackResourceEvents(ctx, []ResourceEventModel{resourceEvent})
+}
+
+func (repo TigrisRepository) TrackResourceEvents(ctx context.Context, models []ResourceEventModel) error {
+	resourceEvents := make([]*ResourceEvent, 0)
+	for _, model := range models {
+		resourceEvents = append(resourceEvents, NewResourceEventFromModel(model))
+	}
+
+	events := tigris.GetCollection[ResourceEvent](repo.db)
+	_, err := events.InsertOrReplace(ctx, resourceEvents...)
+
+	return err
+}
+
+func (repo TigrisRepository) ListResourceEvents(ctx context.Context, args ListResourceEventParams) ([]ResourceEventModel, string, error) {
+	var f []filter.Expr
+
+	if args.Type != "" {
+		f = append(f, filter.Eq("type", args.Type))
+	}
+
+	if args.Source != "" {
+		f = append(f, filter.Eq("source", args.Source))
+	}
+
+	if args.ResourceType != "" {
+		f = append(f, filter.Eq("resourceType", args.ResourceType))
+	}
+
+	if args.ResourceId != "" {
+		f = append(f, filter.Eq("resourceId", args.ResourceId))
+	}
+
+	if args.LastId != "" {
+		lastIdSpec, err := stringToLastIdSpec(args.LastId)
+		if err != nil {
+			return nil, "", service.NewInvalidParameterError("lastId", "")
+		}
+
+		f = append(f, filter.Lt("createdAt", lastIdSpec.CreatedAt))
+		f = append(f, filter.Lt("id", lastIdSpec.ID))
+	}
+
+	f = append(f, filter.Gte("createdAt", args.Since))
+	f = append(f, filter.Lt("createdAt", args.Until))
+
+	//query = fmt.Sprintf("%s %s ORDER BY createdAt DESC, BIN_TO_UUID(id) DESC LIMIT ?", query, strings.Join(conditions, " AND "))
+
+	events := tigris.GetCollection[ResourceEvent](repo.db)
+	it, err := events.ReadWithOptions(ctx, filter.And(f...), fields.All,
```

Comment:
Do you need a deferred call to `it.Close()` below?

Reply:
Good catch! Fixed.